diff --git a/include/flatbuffers/code_generators.h b/include/flatbuffers/code_generators.h
index bccd9fe0..27c08faa 100644
--- a/include/flatbuffers/code_generators.h
+++ b/include/flatbuffers/code_generators.h
@@ -90,6 +90,7 @@ class CodeWriter {
 class BaseGenerator {
  public:
   virtual bool generate() = 0;
+  virtual std::string generateCode();
 
   static std::string NamespaceDir(const Parser &parser, const std::string &path,
                                   const Namespace &ns);
diff --git a/include/flatbuffers/idl.h b/include/flatbuffers/idl.h
index 12b2b142..039c4cde 100644
--- a/include/flatbuffers/idl.h
+++ b/include/flatbuffers/idl.h
@@ -511,6 +511,7 @@ struct IDLOptions {
   // Use flexbuffers instead for binary and text generation
   bool use_flexbuffers;
   bool strict_json;
+  bool ignore_null_scalar;
   bool skip_js_exports;
   bool use_goog_js_export_format;
   bool use_ES6_js_export_format;
@@ -561,6 +562,7 @@ struct IDLOptions {
   std::string proto_namespace_suffix;
   std::string filename_suffix;
   std::string filename_extension;
+  std::string js_ts_global_prefix;
 
   // Possible options for the more general generator below.
   enum Language {
@@ -605,6 +607,7 @@ struct IDLOptions {
   IDLOptions()
       : use_flexbuffers(false),
         strict_json(false),
+        ignore_null_scalar(false),
         skip_js_exports(false),
         use_goog_js_export_format(false),
         use_ES6_js_export_format(false),
@@ -660,6 +663,9 @@ struct ParserState {
         line_start_(nullptr),
         line_(0),
         token_(-1),
+        has_length_(false),
+        len_(0),
+        eof_(nullptr),
         attr_is_trivial_ascii_string_(true) {}
 
  protected:
@@ -683,6 +689,9 @@ struct ParserState {
   const char *line_start_;
   int line_;  // the current line being parsed
   int token_;
+  bool has_length_;
+  unsigned int len_;
+  const char *eof_;
 
   // Flag: text in attribute_ is true ASCII string without escape
   // sequences. Only printable ASCII (without [\t\r\n]).
@@ -801,6 +810,9 @@ class Parser : public ParserState {
   bool Parse(const char *_source, const char **include_paths = nullptr,
              const char *source_filename = nullptr);
 
+  // Set the maximum size to process.
+  void SetContentLength(const bool has_length, const unsigned int len);
+
   // Set the root type. May override the one set in the schema.
   bool SetRootType(const char *name);
 
@@ -1016,6 +1028,10 @@ extern bool GenerateDart(const Parser &parser, const std::string &path,
 extern bool GenerateJava(const Parser &parser, const std::string &path,
                          const std::string &file_name);
 
+// Generate JavaScript or TypeScript code from the definitions in the Parser
+// object. See idl_gen_js.
+extern std::string GenerateJSTSCode(const Parser &parser);
+
 // Generate JavaScript or TypeScript code from the definitions in the Parser
 // object. See idl_gen_js.
 extern bool GenerateJSTS(const Parser &parser, const std::string &path,
diff --git a/src/code_generators.cpp b/src/code_generators.cpp
index 46d65f7c..51ead386 100644
--- a/src/code_generators.cpp
+++ b/src/code_generators.cpp
@@ -82,6 +82,10 @@ const char *BaseGenerator::FlatBuffersGeneratedWarning() {
          " do not modify";
 }
 
+std::string BaseGenerator::generateCode() {
+    return NULL;
+}
+
 std::string BaseGenerator::NamespaceDir(const Parser &parser,
                                         const std::string &path,
                                         const Namespace &ns) {
diff --git a/src/flatc.cpp b/src/flatc.cpp
index 5cb2a80c..734c4047 100644
--- a/src/flatc.cpp
+++ b/src/flatc.cpp
@@ -80,6 +80,7 @@ std::string FlatCompiler::GetUsageString(const char *program_name) const {
     "  --version              Print the version number of flatc and exit.\n"
     "  --strict-json          Strict JSON: field names must be / will be quoted,\n"
     "                         no trailing commas in tables/vectors.\n"
+    "  --ignore-null-scalar Ignore null values for scalar types\n"
     "  --allow-non-utf8       Pass non-UTF-8 input through parser and emit nonstandard\n"
     "                         \\x escapes in JSON. (Default is to raise parse error on\n"
     "                         non-UTF-8 input.)\n"
@@ -232,6 +233,8 @@ int FlatCompiler::Compile(int argc, const char **argv) {
         opts.keep_include_path = true;
       } else if (arg == "--strict-json") {
         opts.strict_json = true;
+      } else if (arg == "--ignore-null-scalar") {
+        opts.ignore_null_scalar = true;
       } else if (arg == "--allow-non-utf8") {
         opts.allow_non_utf8 = true;
       } else if (arg == "--natural-utf8") {
diff --git a/src/idl_gen_js_ts.cpp b/src/idl_gen_js_ts.cpp
index 030c5526..6745ae4c 100644
--- a/src/idl_gen_js_ts.cpp
+++ b/src/idl_gen_js_ts.cpp
@@ -75,7 +75,7 @@ class JsTsGenerator : public BaseGenerator {
         lang_(GetJsLangParams(parser_.opts.lang)) {}
   // Iterate through all definitions we haven't generate code for (enums,
   // structs, and tables) and output them to a single file.
-  bool generate() {
+  std::string generateCode() {
     imported_fileset imported_files;
     reexport_map reexports;
 
@@ -105,6 +105,11 @@ class JsTsGenerator : public BaseGenerator {
       code += exports_code;
     }
 
+    return code;
+  }
+ 
+  bool generate() {
+    std::string code = generateCode();
     return SaveFile(GeneratedFileName(path_, file_name_, parser_.opts).c_str(),
                     code, false);
   }
@@ -223,7 +228,11 @@ class JsTsGenerator : public BaseGenerator {
             exports += "this." + *it + " = " + *it + ";\n";
           }
         }
-        code += *it + " = " + *it + " || {};\n\n";
+        if (parser_.opts.js_ts_global_prefix.empty()) {
+          code += *it + " = " + *it + " || {};\n\n";
+        } else {
+          code += *it + " = " + parser_.opts.js_ts_global_prefix + *it + " || {};\n\n";
+        }
       }
     }
   }
@@ -1377,6 +1386,11 @@ class JsTsGenerator : public BaseGenerator {
 };
 }  // namespace jsts
 
+std::string GenerateJSTSCode(const Parser &parser) {
+    jsts::JsTsGenerator generator(parser, "", "");
+    return generator.generateCode();
+}
+
 bool GenerateJSTS(const Parser &parser, const std::string &path,
                   const std::string &file_name) {
   jsts::JsTsGenerator generator(parser, path, file_name);
diff --git a/src/idl_parser.cpp b/src/idl_parser.cpp
index 28ff3ec0..2d9fc051 100644
--- a/src/idl_parser.cpp
+++ b/src/idl_parser.cpp
@@ -272,6 +272,10 @@ CheckedError Parser::Next() {
   attribute_.clear();
   attr_is_trivial_ascii_string_ = true;
   for (;;) {
+    if (has_length_ && cursor_ >= eof_) {
+      token_ = kTokenEof;
+      return NoError();
+    }
     char c = *cursor_++;
     token_ = c;
     switch (c) {
@@ -1089,7 +1093,7 @@ CheckedError Parser::ParseTable(const StructDef &struct_def, std::string *value,
             ECHECK(SkipAnyJsonValue());
           }
         } else {
-          if (IsIdent("null") && !IsScalar(field->value.type.base_type)) {
+          if (IsIdent("null") && (opts.ignore_null_scalar || !IsScalar(field->value.type.base_type))) {
             ECHECK(Next());  // Ignore this field.
           } else {
             Value val = field->value;
@@ -2358,6 +2362,11 @@ CheckedError Parser::ParseService() {
   return NoError();
 }
 
+void Parser::SetContentLength(const bool has_length, const unsigned int len) {
+  has_length_ = has_length;
+  len_ = len;
+}
+
 bool Parser::SetRootType(const char *name) {
   root_struct_def_ = LookupStruct(name);
   if (!root_struct_def_)
@@ -2822,6 +2831,9 @@ CheckedError Parser::StartParseFile(const char *source,
   file_being_parsed_ = source_filename ? source_filename : "";
   source_ = source;
   ResetState(source_);
+  if (has_length_) {
+    eof_ = source + len_;
+  }
   error_.clear();
   ECHECK(SkipByteOrderMark());
   NEXT();
@@ -2965,6 +2977,8 @@ CheckedError Parser::DoParse(const char *source, const char **include_paths,
         std::string contents;
         if (!LoadFile(filepath.c_str(), true, &contents))
           return Error("unable to load include file: " + name);
+        bool has_length = has_length_;
+        has_length_ = false;
         ECHECK(DoParse(contents.c_str(), include_paths, filepath.c_str(),
                        name.c_str()));
         // We generally do not want to output code for any included files:
@@ -2974,6 +2988,7 @@ CheckedError Parser::DoParse(const char *source, const char **include_paths,
         root_struct_def_ = nullptr;
         file_identifier_.clear();
         file_extension_.clear();
+        has_length_ = has_length;
         // This is the easiest way to continue this file after an include:
         // instead of saving and restoring all the state, we simply start the
         // file anew. This will cause it to encounter the same include
